rre-core/src/main/java/io/sease/rre/core/domain/metrics/impl/NDCGAtK.java

```diff
@@ -0,0 +1,129 @@
+    public ValueFactory createValueFactory(final String version) {
+        return new ValueFactory(this, version) {
+            private BigDecimal dcg = BigDecimal.ZERO;
+
+            @Override
+            public void collect(final Map<String, Object> hit, final int rank, final String version) {
+                if (rank > k) return;
+                judgment(id(hit))
+                        .ifPresent(judgment -> {
+                            final BigDecimal value = gainOrRatingNode(judgment).map(JsonNode::decimalValue).orElse(TWO);
+                            BigDecimal numerator = TWO.pow(value.intValue()).subtract(BigDecimal.ONE);
+                            if (rank == 1) {
+                                dcg = numerator;
+                            } else {
+                                double den = Math.log(rank + 1) / Math.log(2);
+                                dcg = dcg.add(numerator.divide(new BigDecimal(den), 2, RoundingMode.FLOOR));
+                            }
+                        });
+            }
+
+            @Override
+            public BigDecimal value() {
+                if (totalHits == 0) {
+                    return relevantDocuments.size() == 0 ? BigDecimal.ONE : BigDecimal.ZERO;
+                }
+
+                final BigDecimal idealDcg = idealDcg(relevantDocuments);
+                if (dcg.equals(BigDecimal.ZERO) && idealDcg.equals(BigDecimal.ZERO)) {
+                    return BigDecimal.ZERO;
+                }
+
+                return dcg.divide(idealDcg, 2, RoundingMode.FLOOR);
+            }
+        };
+    }
+
+    private BigDecimal idealDcg(final JsonNode relevantDocuments) {
+        final int windowSize = Math.min(relevantDocuments.size(), k);
+        final int[] gains = new int[windowSize];
+
+        final Map<Integer, List<JsonNode>> groups =
+                StreamSupport.stream(relevantDocuments.spliterator(), false)
+                                .collect(groupingBy(doc -> gainOrRatingNode(doc).map(JsonNode::intValue).orElse(2)));
+
+        Set<Integer> ratingValues = groups.keySet();
+        List<Integer> ratingsSorted = new ArrayList<>(ratingValues);
+        ratingsSorted.sort(Collections.reverseOrder());
+        int startIndex = 0;
+        for (Integer ratingValue : ratingsSorted) {
+            if (startIndex < windowSize) {
+                List<JsonNode> docsPerRating = groups.get(ratingValue);
+                int endIndex = startIndex + docsPerRating.size();
+                Arrays.fill(gains, startIndex, Math.min(windowSize, endIndex), ratingValue);
+                startIndex = endIndex;
+            }
+        }
+        
+        BigDecimal result = BigDecimal.ZERO;
+        for (int i = 1; i <= gains.length; i++) {
+            BigDecimal num = TWO.pow(gains[i-1]).subtract(BigDecimal.ONE);
+            double den = Math.log(i + 1) / Math.log(2);
+            result = result.add((num.divide(new BigDecimal(den), 2, RoundingMode.FLOOR)));
```

Comment:
should this 2 be a constant as well? is it the number of decimals?

Reply:
yes